File: v2/golang/client/upload_wait.go
package client

// upload_wait.go — WaitForUpload: polling de upload com tolerância a 404
// transiente (read-your-writes) e semântica IDÊNTICA nos três SDKs.
//
// Paridade: Python `wait_for_upload` / TypeScript `waitForUpload` — mesma
// máquina de estados, mesmos defaults. Qualquer mudança aqui exige o mesmo
// PR nos três SDKs (PARITY_SPEC.md).
//
// Junior Tip [por que 404 vira "pendente" no começo — medido 2026-08-07]:
// as leituras são load-balanced; logo após o POST de upload um follower que
// ainda não aplicou a entrada devolve 404 legítimo por alguns segundos. Antes
// deste helper cada cliente tratava isso de um jeito (o runner Go tolerava, o
// Python morria, o TS passava por sorte de timing). Dentro de NotFoundGrace o
// 404 é espera; DEPOIS dela é erro real — um id inválido não pode virar espera
// infinita (falhar alto, nunca engolir).

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"time"
)

// ErrUploadWaitTimeout is returned by WaitForUpload when the upload did not
// reach a terminal status within Timeout. Use errors.Is to detect it.
var ErrUploadWaitTimeout = errors.New("upload wait timed out")

// WaitUploadOptions configures WaitForUpload. Zero values take the parity
// defaults shared by the three SDKs.
type WaitUploadOptions struct {
	// Timeout is the total wait budget (default 120s).
	Timeout time.Duration
	// Interval is the pause between polls (default 5s).
	Interval time.Duration
	// NotFoundGrace is how long an HTTP 404 is treated as "not applied yet"
	// before becoming a real error (default 30s).
	NotFoundGrace time.Duration
}

// waitUploadDefault* are the parity defaults — keep in lockstep with the
// Python and TypeScript helpers.
const (
	waitUploadDefaultTimeout       = 120 * time.Second
	waitUploadDefaultInterval      = 5 * time.Second
	waitUploadDefaultNotFoundGrace = 30 * time.Second
)

// uploadStatusIsTerminal mirrors the status strings the server emits for a
// finished ingest — success AND failure are both terminal: a failed payload is
// data the caller must inspect, not a transport error.
func uploadStatusIsTerminal(result *UploadStatusResult) bool {
	if result == nil {
		return false
	}
	if result.Completed || result.Error != "" {
		return true
	}
	switch strings.ToLower(result.Status) {
	case "completed", "saved", "done", "failed":
		return true
	}
	return false
}

// WaitForUpload polls UploadStatus until the upload reaches a terminal state.
// It returns the final status payload (including status="failed" — inspect
// it), ErrNotFound when the 404 persists beyond NotFoundGrace, or
// ErrUploadWaitTimeout with the last observed status on budget exhaustion.
func (m *Memory) WaitForUpload(ctx context.Context, uploadID int64, opts WaitUploadOptions) (*UploadStatusResult, error) {
	if opts.Timeout <= 0 {
		opts.Timeout = waitUploadDefaultTimeout
	}
	if opts.Interval <= 0 {
		opts.Interval = waitUploadDefaultInterval
	}
	if opts.NotFoundGrace <= 0 {
		opts.NotFoundGrace = waitUploadDefaultNotFoundGrace
	}

	startedAt := time.Now()
	lastStatus := "never-seen"
	for {
		result, statusErr := m.UploadStatus(ctx, uploadID)
		switch {
		case statusErr == nil:
			if uploadStatusIsTerminal(result) {
				return result, nil
			}
			if result != nil && result.Status != "" {
				lastStatus = result.Status
			}
		case errors.Is(statusErr, ErrNotFound):
			if time.Since(startedAt) >= opts.NotFoundGrace {
				return nil, fmt.Errorf("upload %d still 404 after %s grace: %w",
					uploadID, opts.NotFoundGrace, ErrNotFound)
			}
			lastStatus = "not-found-yet"
		default:
			// Transport/5xx: transient by default — the poll loop IS the retry.
			lastStatus = "error: " + statusErr.Error()
		}

		if time.Since(startedAt)+opts.Interval > opts.Timeout {
			return nil, fmt.Errorf("upload %d not terminal after %s (last=%s): %w",
				uploadID, opts.Timeout, lastStatus, ErrUploadWaitTimeout)
		}
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-time.After(opts.Interval):
		}
	}
}

File: v2/typescript/src/uploadWait.ts
/**
 * AnhurDB TypeScript SDK — the upload-wait STATE MACHINE.
 *
 * One responsibility: given a way to read an upload's status, block until that
 * upload reaches a terminal state, a timeout, or a real error. It lives outside
 * `memory.ts` because it is the only piece of that file with non-trivial
 * temporal logic — three clocks (poll interval, read-your-writes grace, overall
 * deadline) whose interaction is what breaks — and because `memory.ts` is far
 * past the house cut and must not grow to hold it.
 *
 * The status reader is injected rather than imported so the machine can be
 * exercised on its own; the transport contract it depends on is exactly one
 * thing — that a missing upload arrives as an `AnhurError` carrying
 * `statusCode === 404`.
 */

import { AnhurError, AnhurUploadWaitTimeout } from "./types.js";
import type { UploadStatusResult } from "./types.js";

/** Timings of {@link pollUploadUntilTerminal}. All optional. */
export interface UploadWaitOptions {
  /** Give up after this long. Default 120s. */
  timeoutMs?: number;
  /** Sleep between polls. Default 5s. */
  intervalMs?: number;
  /** How long an HTTP 404 still counts as "not visible yet". Default 30s. */
  notFoundGraceMs?: number;
}

/** Statuses the server uses to say "this upload will not change again". */
const TERMINAL_STATUSES = ["completed", "saved", "done", "failed"];

/**
 * Poll `readStatus` until the upload reaches a terminal state.
 *
 * Junior Tip [por que 404 vira "pendente" no começo — medido 2026-08-07]: as
 * leituras do AnhurDB são load-balanced; logo após o POST de upload um follower
 * que ainda não aplicou a entrada devolve 404 legítimo por alguns segundos
 * (read-your-writes). Dentro de `notFoundGraceMs` o 404 é espera; DEPOIS dela
 * ele re-lança — um id inválido não pode virar espera infinita (falhar alto,
 * nunca engolir).
 *
 * Junior Tip [this grace depends on `statusCode`, not on the message]: the test
 * below is `thrown.statusCode === 404`. An error built without its status
 * classifies as `"transport"` and never matches, so the whole grace window
 * silently stops existing and a merely-not-yet-replicated upload is reported as
 * a hard failure. That is why `clientResponse.ts` refuses to construct an HTTP
 * error without its status — this function is the consumer that pays for it.
 *
 * Returns the final payload — INCLUDING `status="failed"` (a failed ingest is
 * terminal data the caller must inspect, not a transport error). Throws the
 * original `AnhurQueryError` when the 404 persists beyond the grace window, and
 * `AnhurUploadWaitTimeout` when no terminal status arrives in time.
 *
 * @param readStatus - Reads the current status (one HTTP GET per call).
 * @param uploadId   - Only used to name the upload in the timeout message.
 * @param options    - Timings; see {@link UploadWaitOptions}.
 */
export async function pollUploadUntilTerminal(
  readStatus: () => Promise<UploadStatusResult>,
  uploadId: number,
  options?: UploadWaitOptions,
): Promise<UploadStatusResult> {
  const timeoutMs = options?.timeoutMs ?? 120_000;
  const intervalMs = options?.intervalMs ?? 5_000;
  const notFoundGraceMs = options?.notFoundGraceMs ?? 30_000;

  const startedAt = Date.now();
  let lastStatus = "never-seen";
  for (;;) {
    try {
      const statusPayload = await readStatus();
      const statusText = String(statusPayload?.status ?? "").toLowerCase();
      // Junior Tip [terminal state comes from `status`, and only from it]: this
      // also tested `Boolean(statusPayload.error)`, a field the handler NEVER
      // emits (see UploadStatusResult). Dead code that reads like a safety net
      // is worse than no net — it invites the next reader to report failures
      // through a channel nobody is listening on.
      if (
        statusPayload?.completed === true ||
        TERMINAL_STATUSES.includes(statusText)
      ) {
        return statusPayload;
      }
      if (statusText) lastStatus = statusText;
    } catch (thrown) {
      const isTransientNotFound =
        thrown instanceof AnhurError &&
        thrown.statusCode === 404 &&
        Date.now() - startedAt < notFoundGraceMs;
      if (!isTransientNotFound) throw thrown;
      lastStatus = "not-found-yet";
    }

    if (Date.now() - startedAt + intervalMs > timeoutMs) {
      throw new AnhurUploadWaitTimeout(
        `upload ${uploadId} not terminal after ${timeoutMs}ms (last=${lastStatus})`,
      );
    }
    await new Promise((resolve) => setTimeout(resolve, intervalMs));
  }
}
